### PR TITLE
[WIP] Handle metadata attribute for declarative mapping of dataset models

### PR DESCRIPTION
**Pull request by jdavcs**
UPDATE: This will work, after all. I'll follow the suggestion in the comments.

ORIGINAL (edited) DESCRIPTION:
~~It appears that mapping HDA and LDDA declaratively may be impossible.~~

Here's the problem. There are 3 models left that are still mapped imperatively. Two of them are `DatasetInstance` classes (HDA and LDDA), which have a `metadata` attribute. BUT. When mapped declaratively, a class is given a `metadata` attribute which [exposes SQLAlchemy's MetaData object](https://docs.sqlalchemy.org/en/13/orm/extensions/declarative/api.html#sqlalchemy.ext.declarative.declarative_base.params.metadata), which is one of the central concepts in SQLAlchemy. Thus, a class cannot be mapped declaratively if it has its own `metadata` attribute. (`sqlalchemy.exc.InvalidRequestError: Attribute name 'metadata' is reserved for the MetaData instance when using a declarative base class.`)

One option is to rename the attribute to, say, `metadata_`. There are *a lot* of "things" in Galaxy that are, or have, metadata, and many of them are not datasets, so there are a lot of occurrences of the term `metadata` in the code base. However, the `.metadata` attribute exposed to tool wrappers needs to remain unchanged (see comment). By setting up a proxy instead of exposing the model directly, we solve both issues: (a) dataset metadata is still available via the `.metadata` attribute + (b) this dramatically cuts down on the number of instances of the term `metadata` that need to be renamed, so the scale of renaming is no longer an issue.

~~While fixing the last 3 unit tests, I came across this: https://github.com/galaxyproject/galaxy/pull/7507. So, if I understand correctly, by renaming the `metadata` attribute we would lose critical backward compatibility.~~ 

The other option that I considered was to leave this as is: imperatively and declaratively mapped models can coexist. It's not as elegant, but it's not a hack either (it's OK as per SQLAlchemy's documentation). But, given the proposed approach (as well as the fact that different mapping leads to models having different base classes, that [may result in different behavior](https://github.com/galaxyproject/galaxy/pull/12064/commits/f8f691434463accf1dea15db6324e24f79490bb4)), this is no longer the optimal solution.

### CODE REVIEW

File: gxupload_0 [hunk 1]
```diff
@@ -0,0 +1,2 @@
+sample data
```
Comment (nsoranzo) [inline gxupload_0:1]: Pretty sure you didn't want to add this.

**Comment by jmchilton** [comment]
Yeah - I think that the thing that is exposed to tool wrappers needs to have ``.metadata`` remain the same - there are a ton of tools in the tool shed.

That said - I think we can pass something to the tools that isn't the models directly right. We already wrap them with proxy objects - just override that proxying to yield the metadata you want instead of the SQL alchemy metadata.

**Comment by jmchilton** [comment]
Also as for that legacy history import test case - I think that shouldn't hold anything back. We're reading data from a JSON to build models - we can do that however we wish.

**Comment by ic4f** [comment]
@jmchilton - That should work! I didn't even think in that direction.. Thanks so much!

File: gxupload_0 (hunk 1 above)
Comment (jdavcs) [reply gxupload_0:1]: Definitely not, thanks!

**Comment by jdavcs** [comment]
Implemented in #13805

**Pull request closed without merging**